gui: Fix regression in *txoutset* in GUI console

hebasto opened the pull request:
The regression was introduced in #19056: if the GUI is running without `-server=1`, the `*txoutset*` call in the console returns "Shutting down".

Fix #19255.

hebasto committed:
gui: Fix regression in GUI console

This change prevents "Shutting down" message during "dumptxoutset",
"gettxoutsetinfo" and "scantxoutset" calls.

MarcoFalke commented on src/rpc/blockchain.cpp:
```diff
@@ -997,7 +997,7 @@ static UniValue gettxoutsetinfo(const JSONRPCRequest& request)
 
     CCoinsView* coins_view = WITH_LOCK(cs_main, return &ChainstateActive().CoinsDB());
-    if (GetUTXOStats(coins_view, stats, RpcInterruptionPoint)) {
+    if (GetUTXOStats(coins_view, stats, IsRPCRunning() ? RpcInterruptionPoint : [] {})) {
```
If the rpc code needs such a workaround for the gui, I'd prefer to at least explain it with a comment

hebasto replied:
Isn't it self-documented: "interrupt RPC only if RPC is running"?

MarcoFalke replied:
Also, it seems racy when the rpc is shut down during flush (which might take a long time)

Maybe the interruption point can be passed in through the `request` similar to how the connman is passed in?

hebasto replied:
> Also, it seems racy when the rpc is shut down during flush (which might take a long time)

Did not get why. `IsRPCRunning()` is checked after `ForceFlushStateToDisk()` call. During GUI shutdown #17659 should ensure the correct `RPCConsole` behavior.

MarcoFalke replied:
* `gettxoutsetinfo`
* `stop` -> Doesn't stop because gettxoutsetinfo is not interruptible
* `gettxoutsetinfo` finishes and server stops

hebasto replied:
[Updated](https://github.com/bitcoin/bitcoin/pull/19323#issuecomment-646482928).

hebasto commented:
Updated 369f5d263ee5a26b62545b604ee0b33f24eaadfa -> 434f74365ebb265bcf04bbb1c87d368b5c6553a4 ([pr19323.01](https://github.com/hebasto/bitcoin/commits/pr19323.01) -> [pr19323.02](https://github.com/hebasto/bitcoin/commits/pr19323.02), [diff](https://github.com/hebasto/bitcoin/compare/pr19323.01..pr19323.02)):

- addressed @MarcoFalke's [comment](https://github.com/bitcoin/bitcoin/pull/19323#discussion_r442483706):
> Also, it seems racy when the rpc is shut down during flush (which might take a long time)
>
>  Maybe the interruption point can be passed in through the `request` similar to how the connman is passed in?

MarcoFalke commented:
weak review ACK 434f74365ebb265bcf04bbb1c87d368b5c6553a4

cc @ryanofsky is the use of `NodeContext` here appropriate?

ryanofsky reviewed (approved):
Code review ACK 434f74365ebb265bcf04bbb1c87d368b5c6553a4 with followup suggestions below.

> weak review ACK [434f743](https://github.com/bitcoin/bitcoin/commit/434f74365ebb265bcf04bbb1c87d368b5c6553a4)
> 
> cc @ryanofsky is the use of `NodeContext` here appropriate?

This seems ok as a workaround but IMO better would be to eliminate `g_rpc_running` global. Might do that by moving g_rpc_running and g_rpcSignals into an RpcContext struct, adding a unique_ptr\<RpcContext> or RpcContext* member to NodeContext, and passing RpcContext& to StartRPC/InterruptRPC/StopRPC.

A more immediate concern with this PR is it that it seems incomplete. I see 4 references to RpcInterruptionPoint that will throw in the GUI, but 434f74365ebb265bcf04bbb1c87d368b5c6553a4 is only fixing one of the four:

https://github.com/bitcoin/bitcoin/blob/5f72ddb7ee4cc177de31f43c69390ee72687222a/src/rpc/blockchain.cpp#L1000
https://github.com/bitcoin/bitcoin/blob/5f72ddb7ee4cc177de31f43c69390ee72687222a/src/rpc/blockchain.cpp#L1979
https://github.com/bitcoin/bitcoin/blob/5f72ddb7ee4cc177de31f43c69390ee72687222a/src/rpc/blockchain.cpp#L2319
https://github.com/bitcoin/bitcoin/blob/5f72ddb7ee4cc177de31f43c69390ee72687222a/src/rpc/blockchain.cpp#L2337

hebasto commented:
Updated 434f74365ebb265bcf04bbb1c87d368b5c6553a4 -> e75b125264e8dc8115fd035b1334b023a4913a11 ([pr19323.02](https://github.com/hebasto/bitcoin/commits/pr19323.02) -> [pr19323.03](https://github.com/hebasto/bitcoin/commits/pr19323.03), [diff](https://github.com/hebasto/bitcoin/compare/pr19323.02..pr19323.03)):

- addressed @ryanofsky's [comment](https://github.com/bitcoin/bitcoin/pull/19323#pullrequestreview-434066791):
> A more immediate concern with this PR is it that it seems incomplete. I see 4 references to RpcInterruptionPoint that will throw in the GUI, but 434f743 is only fixing one of the four

Re: https://github.com/bitcoin/bitcoin/pull/19323#pullrequestreview-434066791
> This seems ok as a workaround but IMO better would be to eliminate `g_rpc_running` global. Might do that by moving g_rpc_running and g_rpcSignals into an RpcContext struct, adding a unique_ptr<RpcContext> or RpcContext* member to NodeContext, and passing RpcContext& to StartRPC/InterruptRPC/StopRPC.

These changes appear non-trivial as `g_rpc_running` is used in `IsRPCRunning()`, and its usage is too broad.

MarcoFalke commented on src/rpc/blockchain.cpp:
```diff
@@ -2143,7 +2147,8 @@ UniValue scantxoutset(const JSONRPCRequest& request)
         }
-        bool res = FindScriptPubKey(g_scan_progress, g_should_abort_scan, count, pcursor.get(), needles, coins);
+        NodeContext& node = EnsureNodeContext(request.context);
+        bool res = FindScriptPubKey(g_scan_progress, g_should_abort_scan, count, pcursor.get(), needles, coins, node);
```
nit: why not pass in the `rpc_interruption_point` like below  for `GetUTXOStats`?

```suggestion
        bool res = FindScriptPubKey(g_scan_progress, g_should_abort_scan, count, pcursor.get(), needles, coins, node.rpc_interruption_point);
```

hebasto replied:
[Updated](https://github.com/bitcoin/bitcoin/pull/19323#issuecomment-646980904).

hebasto commented:
Updated e75b125264e8dc8115fd035b1334b023a4913a11 -> 44ef1170cf1466bfa3ea296c78b821605caa0311 ([pr19323.03](https://github.com/hebasto/bitcoin/commits/pr19323.03) -> [pr19323.04](https://github.com/hebasto/bitcoin/commits/pr19323.04), [diff](https://github.com/hebasto/bitcoin/compare/pr19323.03..pr19323.04)):

- addressed @MarcoFalke's [comment](https://github.com/bitcoin/bitcoin/pull/19323#discussion_r443028301):

> nit: why not pass in the `rpc_interruption_point` like below for `GetUTXOStats`?

ryanofsky reviewed (approved):
Code review ACK 44ef1170cf1466bfa3ea296c78b821605caa0311. Only change since last review is replacing more uses of RpcInterruptionPoint that could also throw and fail unexpectedly.

I think there is room to improve this in the future. RPC globals be eliminated as described https://github.com/bitcoin/bitcoin/pull/19323#pullrequestreview-434066791. Also more importantly these methods should always be interrupted when the gui is shutting down, whether or not -server=0 or -server=1 is used, not just when -server=1 is used.

MarcoFalke commented:
Review ACK 44ef1170cf1466bfa3ea296c78b821605caa0311 , checked that all RpcInterruptionPoint are replaced by node.rpc_interruption_point 👥

<details><summary>Show signature and timestamp</summary>

Signature:

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA512

Review ACK 44ef1170cf1466bfa3ea296c78b821605caa0311 , checked that all RpcInterruptionPoint are replaced by node.rpc_interruption_point 👥
-----BEGIN PGP SIGNATURE-----

iQGzBAEBCgAdFiEE+rVPoUahrI9sLGYTzit1aX5ppUgFAlwqrYAACgkQzit1aX5p
pUiyEgwAzwkY2V+cNbtFvl4G/65/9I3dhHDE68NmHRgq6qQxP0/oA4MKcUrSjaoO
naJMn9LP4L8NMmNR3VQe9lhxu63qzhymxRYCzJJRkXHAWQ1WQMapMsR1NN1861VG
c3Zz0/t6okQHx1o6N5XwMx1YiecPcMEs8h61JGXLq1f17ECVRI8z5oSo4nL3j+FH
oeS4iIszG46CSWlueqZE7Vq7h7J1aEVj3CidHIPsopYRs/WhlvkQDAIkoxqdbWIU
xpk/wQZzUWOFL9C0dhMVybvouJW3ZPuw+AnS11nPBkagmw7DLUZ13+u4kon2qi6a
UNJA670fqlR0mEK6scKy3/PoRJCPl2z05iyB4Kcxv7/FISPQOc9MJ1IZ/fuyikn7
9XsTITd5C+7a3KUEaROe94tEsTOJJ+ZrvAgzNQ93coy+f7LnPyMOLTGVUPagmNpY
B8lpF7manTmkKjuzCHKUXzTqq0505eXYvGP7L+kGmiPxA69ajEbBGgdqKvPv+/NC
ntWM9mto
=KOfS
-----END PGP SIGNATURE-----
```

[Timestamp of file with hash `30e3c19b4b83028940a62fea9a5feb72bad494a01063d4dacab284af3dfd0f8a  -`](https://opentimestamps.org/info/?004f70656e54696d657374616d7073000050726f6f6600bf89e2e884e89294010830e3c19b4b83028940a62fea9a5feb72bad494a01063d4dacab284af3dfd0f8af010583c3095dfa7b646b953f01cb5f85e7a08fff0100e7f3dbabef78f2d37546e9865093a1e08f1045efe15dbf0085e48a665adff5bf20083dfe30d2ef90c8e2c2b68747470733a2f2f626f622e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267fff0107afc18d456c8fd3319a3d57a9fa8387008f1045efe15daf008ed22b4f462ab29bc0083dfe30d2ef90c8e2e2d68747470733a2f2f616c6963652e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267fff01099890543caa412bd70c8905a37f1a1db08f1045efe15dbf008211b047a76a75df10083dfe30d2ef90c8e292868747470733a2f2f66696e6e65792e63616c656e6461722e657465726e69747977616c6c2e636f6df010ae37b3b32d2aec66f5aa2952baa94bd708f1045efe15daf008f161c259f4f75d450083dfe30d2ef90c8e232268747470733a2f2f6274632e63616c656e6461722e636174616c6c6178792e636f6d)

</details>

MarcoFalke commented:
cc @fanquake you might be interested in this, based on your bug report #19255

hebasto commented:
Rebased 44ef1170cf1466bfa3ea296c78b821605caa0311 -> 314b49bd50906c03911d2b17a21a34685a60b3c8 ([pr19323.04](https://github.com/hebasto/bitcoin/commits/pr19323.04) -> [pr19323.05](https://github.com/hebasto/bitcoin/commits/pr19323.05)) due to the conflict with #19328.

ryanofsky reviewed (approved):
Code review ACK 314b49bd50906c03911d2b17a21a34685a60b3c8. Only change since last review is rebase
